src/order/preorder_hom.lean

```diff
@@ -23,43 +23,178 @@ structure preorder_hom (α β : Type*) [preorder α] [preorder β] :=
+{ to_fun := λ b, ⟨function.const α b, λ _ _ _, le_rfl⟩,
+  monotone' := λ b₁ b₂ h x, h }
+
+@[simp] lemma const_comp (f : α →ₘ β) (c : γ) : (const β c).comp f = const α c := rfl
+
+@[simp] lemma comp_const (γ : Type*) [preorder γ] (f : α →ₘ β) (c : α) :
+  f.comp (const γ c) = const γ (f c) := rfl
+
+/-- Given two bundled monotone maps `f`, `g`, `f.prod g` is the map `x ↦ (f x, g x)` bundled as a
+`preorder_hom`. -/
+@[simps] protected def prod (f : α →ₘ β) (g : α →ₘ γ) : α →ₘ (β × γ) :=
+⟨λ x, (f x, g x), λ x y h, ⟨f.mono h, g.mono h⟩⟩
+
+@[mono] lemma prod_mono {f₁ f₂ : α →ₘ β} (hf : f₁ ≤ f₂) {g₁ g₂ : α →ₘ γ} (hg : g₁ ≤ g₂) :
+  f₁.prod g₁ ≤ f₂.prod g₂ :=
+λ x, prod.le_def.2 ⟨hf _, hg _⟩
+
+lemma comp_prod_comp_same (f₁ f₂ : β →ₘ γ) (g : α →ₘ β) :
+  (f₁.comp g).prod (f₂.comp g) = (f₁.prod f₂).comp g :=
+rfl
+
+/-- Given two bundled monotone maps `f`, `g`, `f.prod g` is the map `x ↦ (f x, g x)` bundled as a
+`preorder_hom`. This is a fully bundled version. -/
+@[simps] def prodₘ : (α →ₘ β) →ₘ (α →ₘ γ) →ₘ α →ₘ β × γ :=
+curry ⟨λ f : (α →ₘ β) × (α →ₘ γ), f.1.prod f.2, λ f₁ f₂ h, prod_mono h.1 h.2⟩
+
+/-- Diagonal embedding of `α` into `α × α` as a `preorder_hom`. -/
+@[simps] def diag : α →ₘ α × α := id.prod id
+
+/-- Restriction of `f : α →ₘ α →ₘ β` to the diagonal. -/
+@[simps {simp_rhs := tt}] def on_diag (f : α →ₘ α →ₘ β) : α →ₘ β := (curry.symm f).comp diag
+
+/-- `prod.fst` as a `preorder_hom`. -/
+@[simps] def fst : α × β →ₘ α := ⟨prod.fst, λ x y h, h.1⟩
+
+/-- `prod.snd` as a `preorder_hom`. -/
+@[simps] def snd : α × β →ₘ β := ⟨prod.snd, λ x y h, h.2⟩
+
+@[simp] lemma fst_prod_snd : (fst : α × β →ₘ α).prod snd = id :=
+by { ext ⟨x, y⟩ : 2, refl }
+
+@[simp] lemma fst_comp_prod (f : α →ₘ β) (g : α →ₘ γ) : fst.comp (f.prod g) = f := ext _ _ rfl
+
+@[simp] lemma snd_comp_prod (f : α →ₘ β) (g : α →ₘ γ) : snd.comp (f.prod g) = g := ext _ _ rfl
+
+/-- Order isomorphism between the space of monotone maps to `β × γ` and the product of the spaces
+of monotone maps to `β` and `γ`. -/
+@[simps] def prod_iso : (α →ₘ β × γ) ≃o (α →ₘ β) × (α →ₘ γ) :=
+{ to_fun := λ f, (fst.comp f, snd.comp f),
+  inv_fun := λ f, f.1.prod f.2,
+  left_inv := λ f, by ext; refl,
+  right_inv := λ f, by ext; refl,
+  map_rel_iff' := λ f g, forall_and_distrib.symm }
+
+/-- `prod.map` of two `preorder_hom`s as a `preorder_hom`. -/
+@[simps] def prod_map (f : α →ₘ β) (g : γ →ₘ δ) : α × γ →ₘ β × δ :=
+⟨prod.map f g, λ x y h, ⟨f.mono h.1, g.mono h.2⟩⟩
+
+variables {ι : Type*} {π : ι → Type*} [Π i, preorder (π i)]
+
+/-- Evaluation of an unbundled function at a point (`function.eval`) as a `preorder_hom`. -/
+@[simps {fully_applied := ff}]
+def _root_.pi.eval_preorder_hom (i : ι) : (Π j, π j) →ₘ π i :=
+⟨function.eval i, function.monotone_eval i⟩
+
+/-- The "forgetful functor" from `α →ₘ β` to `α → β` that takes the underlying function,
+is monotone. -/
+@[simps {fully_applied := ff}] def to_fun_hom : (α →ₘ β) →ₘ (α → β) :=
```

Comment:
Perhaps:
```suggestion
@[simps {fully_applied := ff}] def coe_fn_hom : (α →ₘ β) →ₘ (α → β) :=
```
to match `add_monoid_hom.coe_fn_add_monoid_hom`

Reply:
Done. Could you please have a look at the module docstring?